### ISSUE TITLE
'enter' stage not called in some cases

### ISSUE THREAD

**Issue by Illthiriel** [bevyengine/bevy#1117]
**Bevy version**

bevy 0.4.0

**Operating system & version**

Windows 10

**What you did**

The following code allow to reproduce the issue. If I uncomment the `add_stage_before` line, the `enter_system` is not called anymore.

``` rust
use bevy::prelude::*;

#[derive(Clone)]
pub enum AppState {
    Loading,
}

fn main() {
    App::build()
        .add_plugins(DefaultPlugins)
        .add_resource(State::new(AppState::Loading))
        // uncomment the next line and 'enter_system' is not called
        // .add_stage_before(stage::UPDATE, "STAGE_1", StateStage::<AppState>::default())
        .add_stage_after(stage::UPDATE, "STAGE_2", StateStage::<AppState>::default()
            .with_enter_stage(
                AppState::Loading,
                SystemStage::parallel()
                    .with_system(enter_system.system()),
            )
            .with_update_stage(
                AppState::Loading,
                SystemStage::parallel()
                    .with_system(update_system.system()),
            ))
        .run();
}

pub fn enter_system() {
    println!("enter_system");
}

pub fn update_system() {
    println!("update_system");
}
```

**What you expected to happen**

The 'enter' should be called.

**What actually happened**

The 'enter' is not called.

**Comment by TheNeikos** [comment]
The same issue seems to happen with `exit`:

```
use bevy::prelude::*;

#[derive(Clone)]
pub enum AppState {
    Loading,
    Loaded,
}

fn main() {
    App::build()
        .add_plugins(DefaultPlugins)
        .add_resource(State::new(AppState::Loading))
        // uncomment the next line and 'enter_system' is not called
        // .add_stage_before(stage::UPDATE, "STAGE_1", StateStage::<AppState>::default())
        .add_stage_after(
            stage::UPDATE,
            "STAGE_2",
            StateStage::<AppState>::default()
                .with_enter_stage(
                    AppState::Loading,
                    SystemStage::single(enter_system.system()),
                )
                .with_update_stage(
                    AppState::Loading,
                    SystemStage::single(update_system.system()),
                )
                .with_exit_stage(AppState::Loading, SystemStage::single(exit_system.system())),
        )
        .run();
}

pub fn enter_system() {
    println!("enter_system");
}

pub fn update_system(mut state: ResMut<State<AppState>>) {
    println!("update_system");
    state.set_next(AppState::Loaded).unwrap();
}

pub fn exit_system() {
    println!("exit_system");
}
```

**Comment by TheRawMeatball** [comment]
This should be resolved by #1424

**Comment by cart** [comment]
Closing as #1424 is merged
